cpp/rng/rng.cpp

```diff
@@ -0,0 +1,224 @@
+	}
+
+	constexpr uint64_t default_seed = 0x0123456789abcdef;
+	constexpr size_t uint64_s = sizeof(uint64_t);
+	const char *data = static_cast<const char *>(dat);
+
+	if (count > uint64_s) {
+		init_hash(state, from_cptr(data));
+		hash_bytes(state, data, count - uint64_s);
+	}
+	else {
+		uint64_t init_seed = 0;
+		memcpy(&init_seed, data, count);
+		init_hash(state, init_seed);
+	}
+
+	if (!(state[0] || state[1])) {
+		state[0] = default_seed;
+	}
+
+	// This helps prevent low-zero states caused by a bad seed
+	this->discard(discard_on_seed);
+}
+
+// actually does the number crunching and state updates
+// http://en.wikipedia.org/wiki/Xorshift
+static inline uint64_t do_rng(uint64_t &s0, uint64_t &s1) {
+	s1 ^= s1 << 23;
+	s1 = (s1 ^ s0 ^ (s1 >> 17) ^ (s0 >> 26));
+	return s1 + s0;
+}
+
+uint64_t rng::random() {
+	std::swap(this->state[0], this->state[1]);
+	return do_rng(this->state[0], this->state[1]);
+}
+
+// This provides serious optimizations over the standard generator by
+// eliminating memory writes and data dependencies
+template<class T, class Lambda>
+static inline void act_fill(T *dat, uint64_t *state,
+                            size_t len, const Lambda &op) {
+
+	uint64_t s0 = state[0];
+	uint64_t s1 = state[1];
+	size_t ev_len = len - (len & 1);
+
+	for (size_t i = 0; i < ev_len; i += 2) {
+		op(do_rng(s1, s0), dat, i);
+		op(do_rng(s0, s1), dat, i + 1);
+	}
+	if (len & 1) {
+		op(do_rng(s1, s0), dat, len-1);
+		std::swap(s1, s0);
+	}
+	state[0] = s0;
+	state[1] = s1;
+}
+
+void rng::fill(uint64_t *dat, size_t len) {
+	act_fill<>(
+		dat, this->state, len,
+		[](uint64_t v, uint64_t *d, size_t i) {
+			d[i] = v;
+		}
+	);
+}
+
+void rng::discard(size_t len) {
+	act_fill<uint64_t>(
+		nullptr, this->state, len,
+		[](uint64_t, uint64_t *, size_t) {}
+	);
+}
+
+void rng::fill_real(double *dat, size_t len) {
+	act_fill(
+		dat, this->state, len,
+		[](double v, double *d, size_t i) {
+			d[i] = v / rng::max();
+		}
+	);
+}
+
+std::ostream &rng::to_stream(std::ostream &out) const {
+	if ((out << this->state[0] << " ").good() &&
+	    !(out << this->state[1]).fail()) {
+		return out;
+	}
+	throw util::Error(MSG(err) << "Unable to write rng state serialization to output stream");
+}
+
+std::istream &rng::from_stream(std::istream &in) {
+	if ((in >> this->state[0]).good() &&
+	    !(in >> this->state[1]).fail()) {
+		return in;
+	}
+	throw util::Error(MSG(err) << "Unable to read rng state serialization from input stream");
+}
+
+std::string rng::to_string() const {
+	std::stringstream strstr;
+	return (strstr << *this).str();
+}
+
+void rng::from_string(const std::string &instr) {
+	std::stringstream strstr(instr);
+	strstr >> *this;
+}
+
+static uint64_t try_random_device() {
+	uint64_t rand1 = 0;
+	// std::random_device constructor may throw if it isn't available
+	try {
+		std::random_device rnd;
+		if (rnd.entropy() > 0){
+			rand1 = rnd() + (uint64_t(rnd()) << 32);
+		}
+	}
```

Comment:
if you want your comments auto-outdate, you need to make it after the change you want.
The comment is associated with the 15 lines above it i think.